### PR TITLE
[SYCL][E2E] Add `-Werror` flag to sycl e2e tests

### PR DESCRIPTION
**Pull request by ayylol**
This patch adds the `-Werror` flag to all SYCL e2e tests to stop the introduction of new warnings.

Added `-Wno-error=` to existing tests that have warnings (Or made changes to resolve the warnings).

### CODE REVIEW

File: sycl/test-e2e/Regression/check_vector_of_opencl_event.cpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // REQUIRES: opencl
-// RUN: %{build} -o %t.out
+// RUN: %{build} -std=c++20 -o %t.out
```
Comment (maarquitos14) [inline sycl/test-e2e/Regression/check_vector_of_opencl_event.cpp:2]: Why do we need this?

**Review by maarquitos14 — COMMENTED** [review]
Most of the tests have `-Wno-error=...`. Can we try and fix any of these errors? Maybe in separate PRs?

File: sycl/test-e2e/Regression/check_vector_of_opencl_event.cpp (hunk 1 above)
Comment (ayylol) [reply sycl/test-e2e/Regression/check_vector_of_opencl_event.cpp:2]: This warning is generated without it:
``` 
# .---command stderr------------
# | /localdisk2/dgarciao/llvm/sycl/test-e2e/Regression/check_vector_of_opencl_event.cpp:21:38: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++20 extension [-Wc++20-extensions]
# |    21 |   std::vector<cl_event> ClEventVec = get_native<sycl::backend::opencl>(event);
# |       |                                      ^
# | 1 warning generated.
# | /localdisk2/dgarciao/llvm/sycl/test-e2e/Regression/check_vector_of_opencl_event.cpp:21:38: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++20 extension [-Wc++20-extensions]
# |    21 |   std::vector<cl_event> ClEventVec = get_native<sycl::backend::opencl>(event);
# |       |                                      ^
# | 1 warning generated.
# \`-----------------------------
```

**Comment by ayylol** [comment]
> Most of the tests have `-Wno-error=...`. Can we try and fix any of these warnings? Maybe in separate PRs?

@maarquitos14 
From what I found 20 of these tests were explicitly testing deprecated functionality (I can give more info on these, if you'd like). The tests in syclcompat mainly have issues because including some syclcompat files creates warnings. For the rest I could take a better look at to see if they could be resolved. However I thought it might be a good idea to add the flag before doing so because it would add a visible marker in the code base of the tests that need work.

**Review by maarquitos14 — APPROVED** [review]

**Comment by maarquitos14** [comment]
> > Most of the tests have `-Wno-error=...`. Can we try and fix any of these warnings? Maybe in separate PRs?
> 
> @maarquitos14 From what I found 20 of these tests were explicitly testing deprecated functionality (I can give more info on these, if you'd like). The tests in syclcompat mainly have issues because including some syclcompat files creates warnings. For the rest I could take a better look at to see if they could be resolved. However I thought it might be a good idea to add the flag before doing so because it would add a visible marker in the code base of the tests that need work.

It's definitely a good idea. On a side note, just wondering if having tests for deprecated functionalities makes sense or those should be dropped.

**Review by sarnex — APPROVED** [review]
tools/esimd LGTM! I agree we should see if we can fix some of the warnings that are not expected as part of the test in a separate PR

**Review by Alcpz — APPROVED** [review]

File: sycl/test-e2e/Regression/check_vector_of_opencl_event.cpp (hunk 1 above)
Comment (bader) [reply sycl/test-e2e/Regression/check_vector_of_opencl_event.cpp:2]: Folks this is the wrong way to fix this warning. The test is not supposed to use C++20 features.

Please replace `get_native<sycl::backend::opencl>(event)` to `sycl::get_native<sycl::backend::opencl>(event)` at line 20.

**Review by AllanZyne — APPROVED** [review]
LGTM on asan part

**Comment by againull** [comment]
@unified-runtime-reviewers Friendly ping.

**Review by pbalcer — APPROVED** [review]

**Pull request merged**